x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -212,6 +226,10 @@ private Store.MetadataSnapshot getMetadata() throws IOException {
         }
 
         private long readFileBytes(String fileName, BytesReference reference) throws IOException {
+            // Ensure no file system traversal is possible
+            if (false == getMetadata().contains(fileName)) {
+                throw new IllegalArgumentException("invalid file name [" + fileName + "]");
+            }
```

Comment:
This logic here as well as the above `ensureSessionShardIdConsistency` method are behaviour change for existing CCR as well. They feel safe enough and are likely useful enhancements. We could have a distributed team member to take a look to be sure. Alternatively we could try conditionally disable these checks for RCS 1 leader clusters. But it may not be easy since this service layer does not have knowledge about the cluster connections.

Reply:
I'm leaning towards not disabling them conditionally since as you say it would be complicated to do this; besides, these seem like reasonable validation steps.

Reply:
One more unrelated sanity check: the filenames are for Lucene-maintained segment info files, with a fixed format of `segment_N`; having poked around the code a bit (e.g. this is the relevant ES-side [part](https://github.com/elastic/elasticsearch/blob/7fb344238dcad615282b70557b70c67eede482ca/server/src/main/java/org/elasticsearch/index/store/Store.java#L789)) I don't think there is a way for someone to maliciously (and manually) add a file with a name that would enable directory traversal, unless perhaps they had access to the file system; in that case though they would not benefit from a directory traversal attack to begin with. Was wondering if this is roughly what you have found as well.

Reply:
Yes basically that. Also, I don't think it is possible to physically add a file with a name that enables directory traversal. The file sytsem won't allow such file name to exist in the first place.